### PR TITLE
adding checks for pinning and private registry secrets

### LINKED ISSUE #6: Add support for digest pinning
The operator should support digest pinning as documented in the Operator Certification CI Pipeline Instructions.

https://github.com/redhat-openshift-ecosystem/certification-releases/blob/main/4.9/ga/ci-pipeline.md#digest-pinning-config

### LINKED ISSUE #7: Add support for private container registries
The operator should support private container registries as documented in the Operator Certification CI Pipeline Instructions.

https://github.com/redhat-openshift-ecosystem/certification-releases/blob/main/4.9/ga/ci-pipeline.md#required-step-will-be-optional-soon---if-using-a-private-container-registry

### PR DESCRIPTION
**Pull request by dcurran90**
Two new functions ensureDockerRegistrySecret and ensureGithubSSHSecret added to check for secrets related to private registry and digest pinning respectively.
resource.go simply logs if the secrets are not present because they are optional

Fixes #6 
Fixes #7

### CODE REVIEW

File: controllers/secret.go
Comment (samira-barouti) [inline controllers/secret.go:185]: I believe Sid's PR to refactor secrets could be used here.

File: controllers/resource.go
Comment (acornett21) [inline controllers/resource.go:72]: We should stick with the existing pattern for these secrets as well. It probably would make sense after the refactor PR lands in main.

**Comment by acornett21** [comment]
For `github-ssh-credentials` we need to insure that `id_rsa` exists and is populated as well. Im not sure if we need/have to do this for each `key` for the docker registry, since it seems that `cmd` the cluster already validates for empties on creation.

File: controllers/resource.go
Comment (bcrochet) [reply controllers/resource.go:72]: What's the point of calling this if we are ignoring the return? Is there some kind of side effect happening in secretExists?

**Review by bcrochet — CHANGES_REQUESTED** [review]

File: controllers/secret.go
Comment (dcurran90) [reply controllers/secret.go:185]: great point. will wait for that PR and refactor

File: controllers/resource.go
Comment (dcurran90) [reply controllers/resource.go:72]: however, i'll wait for the refactor.

Comment (dcurran90) [reply controllers/resource.go]: resolving because updated with refactored cod

File: controllers/secret.go
Comment (samira-barouti) [inline controllers/secret.go:97]: I believe this should be return r.ensureSecret(secretName, defaultGithubSSHSecretKeyName, meta). If ensureSecret throws an error, we don't handle it here.

**Review by samira-barouti — CHANGES_REQUESTED** [review]

File: controllers/secret.go
Comment (dcurran90) [reply controllers/secret.go:97]: so the purpose of the function is just to check if the secret exists and log for the users. The secrets are optional and so it should't block if they didn't create it

File: controllers/secret.go
Comment (dcurran90) [reply controllers/secret.go:112]: secrets are optional. only purpose of the function is to check if they exist and log for the users knowledge

File: controllers/secret.go
Comment (samira-barouti) [reply controllers/secret.go:97]: You can handle ignoring the error in the caller function. And it is consistent with what has been done for other secrets.

File: controllers/secret.go
Comment (bcrochet) [reply controllers/secret.go:112]: I'm with @samira-barouti on this one. If anything, I'd move the logging from the ensureSecret method up to here. You say the secret is optional, but that is not very clear here. Someone else reading the code would need to go looking into the ensureSecret method to find out what this is doing.

**Review by bcrochet — CHANGES_REQUESTED** [review]

**Comment by bcrochet** [comment]
Am I missing something? What in this has to do with pinning?

**Review by bcrochet — APPROVED** [review]
/lgtm

File: controllers/resource.go
Comment (samira-barouti) [inline controllers/resource.go:64]: IMO, using log.Warning might be more appropriate since we are handling an error.

**Review by samira-barouti — COMMENTED** [review]
A nit

**Review by acornett21 — COMMENTED** [review]
Non blocking: Since these new types are optional, do we want to make them as such explicitly? And maybe also include them in the `sample` yaml, so at least operator hub might end up looking/feeling complete. Or are we going to go down the documentation route for these?

I'm fine either way, but felt this might be a good place to discuss.

**Review by bcrochet — CHANGES_REQUESTED** [review]
Since we don't have the checks landed yet, you will need to run 'make generate' and 'make manifests'. Probably 'make bundle' too.

**Comment by dcurran90** [comment]
updated with make generate, manifests, bundle run. waiting for review

File: controllers/resource.go
Comment (dcurran90) [reply controllers/resource.go:64]: hmm....looks like our logger doesn't have warning? weird. looking into it

Comment (dcurran90) [reply controllers/resource.go:64]: yeah, okay. no such thing as warning or debug in this logger. it has V-levels
https://pkg.go.dev/github.com/go-logr/logr#section-readme

I could put like `log.V(2).Info("Docker Registry secret message")`
but I wouldn't say that's necessarily more clear and it would disappear if verbosity is turned down

Comment (acornett21) [reply controllers/resource.go:72]: @dcurran90 is correct, there is no `warn` in the library used by controller runtime. I'm fine with this being an `info` and I think it would be bad user experience to log `error` for something that's optional.

Comment (itroyano) [reply controllers/resource.go:72]: We should still propagate the err itself, unless it's a `errors.IsNotFound(err)`.

Comment (bcrochet) [reply controllers/resource.go:72]: @itroyano has a good point.

**Review by bcrochet — CHANGES_REQUESTED** [review]
There doesn't need to be 3 commits here. 2 maybe... (the initial changes, then the generated stuff).

**Comment by dcurran90** [comment]
updated to two commits

**Review by acornett21 — APPROVED** [review]

**Review by bcrochet — APPROVED** [review]
/lgtm

**Pull request merged**
